src/ch/ch_monitor.c

```diff
@@ -560,6 +564,334 @@ chMonitorBuildSocketCmd(virDomainObjPtr vm, const char *socket_path)
     return cmd;
 }
 
+static const char *virCHMonitorEventStrings[] = {
+    "vmm:starting",
+    "vmm:shutdown",
+    "vm:booting", "vm:booted",
+    "vm:pausing", "vm:paused",
+    "vm:resuming", "vm:resumed",
+    "vm:snapshotting", "vm:snapshotted",
+    "vm:restoring", "vm:restored",
+    "vm:resizing", "vm:resized",
+    "vm:shutdown", "vm:deleted",
+    "cpu_manager:create_vcpu",
+    "virtio-device:activated", "virtio-device:reset"
+};
+
+static virCHMonitorEvent virCHMonitorEventFromString(const char *sourceStr,
+        const char *eventStr)
+{
+    int i;
+    g_auto(virBuffer) buf = VIR_BUFFER_INITIALIZER;
+    const char *event;
+
+    virBufferAsprintf(&buf, "%s:%s", sourceStr, eventStr);
+    event = virBufferCurrentContent(&buf);
+
+    for (i = 0; i < virCHMonitorEventMax; i++) {
+        if STREQ(event, virCHMonitorEventStrings[i])
+            break;
+    }
+
+    return i;
+}
+
+static int virCHMonitorValidateEventsJSON(char *buffer, size_t sz,
+            bool *incomplete)
+{
+    int blocks = 0;
+    int events = 0;
+    int i = 0;
+
+    /*
+     * Check if the message is a wellformed JSON.
+     */
+    while (i < sz) {
+        if (buffer[i] == '{') {
+            blocks++;
+        } else if (buffer[i] == '}'){
+            if (blocks == 0) {
+                VIR_WARN("JSON from monitor is not well-formed!");
+                return -1;
+            }
+            blocks--;
+            if (blocks == 0)
+                events++;
+        }
+        i++;
+    }
+    *incomplete = blocks != 0 ? true : false;
+    VIR_DEBUG("Total events received: %d, incomplete: %d",
+            events, *incomplete);
+
+    return events;
```

Comment:
instead of parsing this manually, can you use `virJSONValueFromString` and `virJSONValueObjectKeysNumber` or `virJSONValueObjectForeachKeyValue` to get the number of events?

Reply:
The reason I did it this way is because, virJSONValueFromString expects a single JSON document. But we may have multiple documents received from monitor-fd due to multiple events being notified in succession. So we need to manually parse and split it into individual documents and then pass it on to virJSONValueFromString. Hope this clarifies

Reply:
how about converting these multiple documents into an array (adding `[` at the beginning and `]` at the end of the string) and parse it using `virJSONValueFromString`?

Reply:
I thought about that and infact had a discussion in a review comment in this PR. The problem is,  the event data from cloud-hypervisor is of streaming nature and currently virjson do not support streaming data(virJSONValueFromString calls yajl_complete_parse() assuming that the string is a complete json). So firstly we would need to support a streaming API in virjson. Trying to wrap in an array would need other complications as well. We would need to think more about it and to come up with a good design to have this feature.